principal/apis/eventstream/eventstream.go

```diff
@@ -225,26 +242,15 @@ func (s *Server) sendFunc(c *client, subs eventstreamapi.EventStream_SubscribeSe
 		return fmt.Errorf("panic: invalid data in sendqueue: want: %T, have %T", cloudevents.Event{}, item)
 	}
 
-	prEv, err := format.ToProto(ev)
-	if err != nil {
-		return fmt.Errorf("panic: could not serialize event to wire format: %v", err)
+	eventWriter, exists := s.eventWriters[c.agentName]
+	if !exists {
+		return fmt.Errorf("panic: event writer not found for agent %s", c.agentName)
```

Comment:
At the moment, it appears that if principal attempts to send a message to agent, before agent has connected, the message will be discarded.

This causes a problem in this case, for example:
1) Principal process stops.
2) User makes changes to one or more Argo CD Application/AppProjects/etc.
3) Principal process starts.
4) Principal process starts calls 'StartInformer' to setup K8s watch and start processing K8s events
5) Principal receives events for all watched resources from informer, e.g. Applications, and attempts to send the new Application state (from step 2) to agent
    - But agent hasn't connected yet, so the events are discarded.
6) Agent connects.
    - But agent missed the initial Application updates, above, so no new messages are sent.

You can see an example of this in `principal/apis/eventstream/eventstream.go` in `sendFunc`, where:
* If there is no send queue (outbox) for an agent (by name), the event is discarded and we stop processing events
* If there IS a send queue, but not an eventWriter for that agent (by name), we likewise discard the event and stop processing events.

AFAICT the solution to this is either:
* A) we keep agent events in the queue, even if the agent hasn't connected yet (don't discard them).
    * The only issue is how to know what the name of the agent is before it connects, and I presume we can get this from the namespace on principal side (in the managed agent case)
* or, B) when an agent first connects (after principal OS process startup), we manually re-queue all events for all resources in the agent namespace on principal (e.g. we cause all the events to be re-reconciled by the informer, which cause messages to be sent containing the contents).

My personal preference is A).

Reply:
Thanks for the detailed explanation. Since there are multiple ways of implementing this, we should probably address it in a separate PR to ensure that we don't miss any corner cases. I think we are also tracking this issue here https://github.com/argoproj-labs/argocd-agent/issues/94. WDYT?